File: edge-engine/src/sqrt_kalman.cpp
#include "driftless/sqrt_kalman.h"

#include "driftless/cholesky_update.h"

#include <cmath>

namespace driftless::linalg {
namespace {

// Chi-squared upper-tail critical values. Only the degrees of freedom this
// engine measures in are tabulated: 1 (map-match cross-track), 2 (NHC lateral
// and vertical), 3 (GNSS position), 6 (GNSS position and velocity).
struct ChiSquaredEntry {
    int degrees_of_freedom;
    double at_95;
    double at_99;
    double at_999;
};

constexpr ChiSquaredEntry kChiSquaredTable[] = {
    {1, 3.841, 6.635, 10.828},
    {2, 5.991, 9.210, 13.816},
    {3, 7.815, 11.345, 16.266},
    {6, 12.592, 16.812, 22.458},
};

}  // namespace

SqrtKalmanUpdate arrayFormUpdate(
    const Eigen::Ref<const Eigen::MatrixXd>& sqrt_covariance,
    const Eigen::Ref<const Eigen::MatrixXd>& H,
    const Eigen::Ref<const Eigen::MatrixXd>& sqrt_R) {
    const Eigen::Index n = sqrt_covariance.rows();
    const Eigen::Index m = H.rows();

    // Pre-array. Triangularizing this is the whole algorithm; the blocks below
    // are then just read off.
    Eigen::MatrixXd pre_array = Eigen::MatrixXd::Zero(n + m, n + m);
    pre_array.topLeftCorner(m, m) = sqrt_R;
    pre_array.topRightCorner(m, n) = H * sqrt_covariance;
    pre_array.bottomRightCorner(n, n) = sqrt_covariance;

    // qrToLowerTriangular(A) returns L with L*L^T = A^T*A, so transposing gives
    // the L with L*L^T = pre_array * pre_array^T, which is what the identities
    // in the header are stated over.
    const Eigen::MatrixXd post_array =
        qrToLowerTriangular(pre_array.transpose());

    SqrtKalmanUpdate result;
    result.sqrt_innovation_covariance = post_array.topLeftCorner(m, m);
    result.sqrt_covariance_posterior = post_array.bottomRightCorner(n, n);

    // K_bar = P*H^T*S_nu^-T sits in the lower-left block, so the gain is one
    // triangular solve away: K = K_bar * S_nu^-1.
    const Eigen::MatrixXd gain_bar = post_array.bottomLeftCorner(n, m);
    result.gain = result.sqrt_innovation_covariance.transpose()
                      .triangularView<Eigen::Upper>()
                      .solve(gain_bar.transpose())
                      .transpose();
    return result;
}

double normalizedInnovationSquared(
    const Eigen::Ref<const Eigen::MatrixXd>& sqrt_innovation_covariance,
    const Eigen::Ref<const Eigen::VectorXd>& innovation) {
    // nu^T (S S^T)^-1 nu == ||S^-1 nu||^2, by forward substitution.
    const Eigen::VectorXd whitened =
        sqrt_innovation_covariance.triangularView<Eigen::Lower>().solve(innovation);
    return whitened.squaredNorm();
}

double chiSquaredThreshold(int degrees_of_freedom, double confidence) {
    for (const auto& entry : kChiSquaredTable) {
        if (entry.degrees_of_freedom != degrees_of_freedom) continue;
        if (std::abs(confidence - 0.95) < 1e-6) return entry.at_95;
        if (std::abs(confidence - 0.99) < 1e-6) return entry.at_99;
        if (std::abs(confidence - 0.999) < 1e-6) return entry.at_999;
        return -1.0;
    }
    return -1.0;
}

}  // namespace driftless::linalg

File: android/app/src/main/java/com/driftless/math/SqrtKalman.kt
package com.driftless.math

import kotlin.math.abs

data class SqrtKalmanUpdate(
    val sqrtInnovationCovariance: Matrix,
    val gain: Matrix,
    val sqrtCovariancePosterior: Matrix,
)

object SqrtKalman {

    private data class ChiSquaredEntry(
        val dof: Int,
        val at95: Double,
        val at99: Double,
        val at999: Double,
    )

    private val chiSquaredTable = arrayOf(
        ChiSquaredEntry(1, 3.841, 6.635, 10.828),
        ChiSquaredEntry(2, 5.991, 9.210, 13.816),
        ChiSquaredEntry(3, 7.815, 11.345, 16.266),
        ChiSquaredEntry(6, 12.592, 16.812, 22.458),
    )

    /**
     * Kailath array form of the linear Kalman update in square-root covariance coordinates.
     *
     * Pre-array:
     *     A = [ sqrt_R   H * S ]   ((m + n) x (m + n))
     *         [   0        S   ]
     *
     * Transposed and triangularized via qrToLowerTriangular:
     *     post_array = [ S_nu    0  ]
     *                  [ K_bar  S+  ]
     *
     * Gain: K = K_bar * S_nu^-1 = (S_nu^T \ K_bar^T)^T
     */
    fun arrayFormUpdate(sqrtCovariance: Matrix, H: Matrix, sqrtR: Matrix): SqrtKalmanUpdate {
        val n = sqrtCovariance.rows
        val m = H.rows

        val preArray = Matrix.zeros(n + m, n + m)
        preArray.setBlock(0, 0, sqrtR)
        preArray.setBlock(0, m, H * sqrtCovariance)
        preArray.setBlock(m, m, sqrtCovariance)

        val postArray = CholeskyUpdate.qrToLowerTriangular(preArray.transpose())

        val sqrtInnovationCovariance = postArray.topLeftCorner(m, m)
        val sqrtCovariancePosterior = postArray.bottomRightCorner(n, n)
        val gainBar = postArray.bottomLeftCorner(n, m)

        // Solve S_nu^T * Y = gainBar^T -> K = Y^T
        val sNuT = sqrtInnovationCovariance.transpose()
        val y = sNuT.solveUpperTriangular(gainBar.transpose())
        val gain = y.transpose()

        return SqrtKalmanUpdate(
            sqrtInnovationCovariance = sqrtInnovationCovariance,
            gain = gain,
            sqrtCovariancePosterior = sqrtCovariancePosterior,
        )
    }

    /**
     * Normalized Innovation Squared: nu^T * (S_nu * S_nu^T)^-1 * nu = ||S_nu^-1 * nu||^2
     */
    fun normalizedInnovationSquared(sqrtInnovationCovariance: Matrix, innovation: DoubleArray): Double {
        val whitened = sqrtInnovationCovariance.solveLowerTriangular(innovation)
        var sum = 0.0
        for (x in whitened) {
            sum += x * x
        }
        return sum
    }

    /**
     * Chi-squared upper-tail critical values.
     */
    fun chiSquaredThreshold(degreesOfFreedom: Int, confidence: Double): Double {
        for (entry in chiSquaredTable) {
            if (entry.dof != degreesOfFreedom) continue
            return when {
                confidence >= 0.999 - 1e-4 -> entry.at999
                confidence >= 0.99 - 1e-4 -> entry.at99
                confidence >= 0.95 - 1e-4 -> entry.at95
                else -> entry.at95
            }
        }
        return -1.0
    }
}
